app/src/main/java/org/wikipedia/analytics/SuggestedEditsFunnel.kt

```diff
@@ -0,0 +1,217 @@
+                "source", invokeSource.getName()
+        )
+    }
+
+    private class SuggestedEditsStatsTypeAdapter : JsonSerializer<SuggestedEditStats> {
+        override fun serialize(src: SuggestedEditStats, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
+            return if (src.isEmpty) JsonNull.INSTANCE else GsonUtil.getDefaultGson().toJsonTree(src, typeOfSrc)
+        }
+    }
+
+    private class SuggestedEditStatsCollection {
+        @SerializedName("a-d")
+        val addDescriptionStats = SuggestedEditStats()
+
+        @SerializedName("t-d")
+        val translateDescriptionStats = SuggestedEditStats()
+
+        @SerializedName("a-c")
+        val addCaptionStats = SuggestedEditStats()
+
+        @SerializedName("t-c")
+        val translateCaptionStats = SuggestedEditStats()
+
+        @SerializedName("i-t")
+        val imageTagStats = SuggestedEditStats()
+    }
+
+    private class SuggestedEditStats {
+        @SerializedName("imp")
+        var impressions = 0
+
+        @SerializedName("clk")
+        var clicks = 0
+
+        @SerializedName("sg")
+        var suggestionsClicked = 0
+
+        @SerializedName("cxl")
+        var cancels = 0
+
+        @SerializedName("suc")
+        var successes = 0
+
+        @SerializedName("fl")
+        var failures = 0
+        val isEmpty: Boolean
+            get() = impressions == 0 && clicks == 0 && suggestionsClicked == 0 && cancels == 0 && successes == 0 && failures == 0
+    }
+
+    companion object {
+        private var INSTANCE: SuggestedEditsFunnel? = null
+        private const val SCHEMA_NAME = "MobileWikiAppSuggestedEdits"
+        private const val REV_ID = 18949003
+        private const val SUGGESTED_EDITS_UI_VERSION = "1.0"
+        private const val SUGGESTED_EDITS_API_VERSION = "1.0"
+        const val SUGGESTED_EDITS_ADD_COMMENT = "#suggestededit-add " + SUGGESTED_EDITS_UI_VERSION
+        const val SUGGESTED_EDITS_TRANSLATE_COMMENT = "#suggestededit-translate " + SUGGESTED_EDITS_UI_VERSION
+        const val SUGGESTED_EDITS_IMAGE_TAG_AUTO_COMMENT = "#suggestededit-imgtag-auto " + SUGGESTED_EDITS_UI_VERSION
+        const val SUGGESTED_EDITS_IMAGE_TAG_CUSTOM_COMMENT = "#suggestededit-imgtag-custom " + SUGGESTED_EDITS_UI_VERSION
+
+        operator fun get(invokeSource: InvokeSource): SuggestedEditsFunnel? {
+            if (INSTANCE == null) {
+                INSTANCE = SuggestedEditsFunnel(WikipediaApp.getInstance(), invokeSource)
+            } else if (INSTANCE!!.invokeSource != invokeSource) {
+                INSTANCE!!.log()
+                INSTANCE = SuggestedEditsFunnel(WikipediaApp.getInstance(), invokeSource)
+            }
+            return INSTANCE
+        }
+
+        fun get(): SuggestedEditsFunnel? {
+            return if (INSTANCE != null && INSTANCE!!.invokeSource != InvokeSource.SUGGESTED_EDITS) {
+                INSTANCE
+            } else Companion[InvokeSource.SUGGESTED_EDITS]
+        }
```

Comment:
Looks like these static functions should return non-null variable.

These can be done by adding `!!` or `as SuggestedEditsFunnel`